### ISSUE TITLE
Training crash with "uncorrectable ECC error" and NCCL timeout?

### ISSUE THREAD

**Issue by 2402227288** [huggingface/trl#5094]
<img width="1632" height="646" alt="Image" src="https://github.com/user-attachments/assets/cbb49b6c-1dfe-4b99-8a5c-8b09f769699b" />
I am experiencing a consistent training hang (NCCL Watchdog Timeout) during Supervised Fine-Tuning (SFT). The timeout always occurs at the exact same fixed step during the training loop. I am using trl==0.18.0.

I have tried multiple DeepSpeed ZeRO configurations, but none of them resolved the issue. Here is a summary of my experiments:

ZeRO-2: Results in OOM (Out of Memory) in the middle of training.

ZeRO-2 with Offload: Hangs at a fixed step, eventually triggering the NCCL Watchdog Timeout.

ZeRO-3: Gets stuck at the very beginning. The progress bar cannot even move from step 0 to step 1.

Data Context (Possible Cause):
My dataset includes very long videos, which require extracting around 700 frames per sample. I suspect that the massive sequence length or the I/O bottleneck during data loading might be related to the training stall.

Key Observation / Contrast:
What confuses me the most is that when I run GRPO training using ZeRO-3 on the exact same setup, it works perfectly fine without encountering any OOM or timeout issues. The problem seems to be exclusively tied to the SFT pipeline.

Environment:
trl version: 0.18.0

Task: SFT (Vision-Language Model)

Distributed backend: NCCL / DeepSpeed

Could you please provide some insights into why SFT behaves so differently from GRPO in this context, or if there is any known issue regarding SFT data collation/padding with extremely long multimodal sequences under ZeRO-3? Any advice on how to bypass this would be greatly appreciated!

**Comment by ShuimuChen-hyq** [comment]
```


import logging
import os
import sys

import torch
import datetime  # 必须引入这个

# 找到原来调用 init_process_group 的地方，修改如下：
torch.distributed.init_process_group(
    backend="nccl", 
    # 将默认的 30 分钟 (1800s) 改为 3 小时，甚至更长
    timeout=datetime.timedelta(seconds=10800),  
    # ... 其他参数保持不变
)


import datasets
from dataclasses import dataclass, field
from typing import Optional
import transformers
from datasets import load_dataset, load_from_disk, Dataset, DatasetDict
from transformers import AutoTokenizer, set_seed, AutoProcessor
from transformers.trainer_utils import get_last_checkpoint
import trl
from trl import (
    ModelConfig,
    SFTConfig,
    ScriptArguments,
    SFTTrainer,
    TrlParser,
    get_kbit_device_map,
    get_peft_config,
    get_quantization_config,
)

from tqdm import tqdm
import json
import random

from time_r1.utils.qwen_vl_utils import process_vision_info
from time_r1.dataset.lazy_dataset_sft import LazyVLDatasetSFT
logger = logging.getLogger(__name__)
import re
from copy import deepcopy
from functools import partial



@dataclass
class SFTScriptArguments(ScriptArguments):
    """
    Script arguments for the GRPO training script.

    Args:
        reward_funcs (`list[str]`):
            List of reward functions. Possible values: 'iou', 'format'.
    """
    tool_name_list: list[str] = field(
        default_factory=lambda: ["seek_video_frames"],
        metadata={"help": "List of video tool names."},
    )
    train_data_path: str = field(
        default=None,
        metadata={"help": "Path to the training data JSON file."},
    )
    eval_data_path: str = field(
        default=None,
        metadata={"help": "Path to the evaluation data JSON file."},
    )

    video_folder: str = field(
        default="workdir/datasets",  # Replace with your actual video folder path
        metadata={"help": "Path to the folder containing video files."},
    )

    prompt_template: str = field(
        default="v3",
        metadata={"help": "Prompt template to use."},
    )
    total_video_tokens: int = field(
        default=15360,
        metadata={"help": "Maximum number of video tokens."},
    )
    
    max_per_frame_tokens: int = field(
        default=192,
        metadata={"help": "Maximum number of frame tokens."},
    )
    
    min_per_frame_tokens: int = field(
        default=16,
        metadata={"help": "Minimum number of frame tokens."},
    )
    
    max_frames: int = field(
        default=768,
        metadata={"help": "Maximum number of frames to preview."},
    )

    append_time_instruction: bool = field(
        default=False,
        metadata={"help": "Whether to append time instruction."},
    )

def compute_tool_response_mask(seq, id_im_start=151644, id_tool=14172):
    """
    向量化计算掩码：从 seq (1, T) 形状的 LongTensor 中屏蔽所有 `<|im_start|>tool…<|im_end|>` 区域。
    返回长度 T 的 IntTensor，其中 1 表示保留，0 表示屏蔽。
    这里 id_im_start 和 id_tool 与模型相关，需要从 tokenizer 中获取。
    id_im_start = tokenizer.convert_tokens_to_ids("<|im_start|>")
    id_tool     = tokenizer.convert_tokens_to_ids("tool")
    """
    # 1) 标记所有 "<|im_start|>"
    is_im_start = seq == id_im_start
    # 2) 计算 region_id
    region_id = is_im_start.int().cumsum(dim=1)
    # 3) 标记段开头是否为工具段
    next_is_tool = torch.zeros_like(seq, dtype=torch.bool)
    next_is_tool[:, :-1] = is_im_start[:, :-1] & (seq[:, 1:] == id_tool)
    # 4) 累加得到每段是否为工具段
    region_flag = torch.zeros_like(region_id)
    region_flag = region_flag.scatter_add(
        dim=1,
        index=region_id,
        src=next_is_tool.int().to(region_flag.dtype)
    )
    # 5) 映射回每个 token 是否在工具段
    tool_region_mask = region_flag.gather(dim=1, index=region_id)
    # 6) 最终掩码：非工具段为 1，工具段为 0
    completion_mask = (~tool_region_mask.bool()).int()
    return completion_mask # 最终工具段由于为0被屏蔽，非工具段由于为1被保留

processor = None

def clear_output(text, reserve_pad=True):
    if reserve_pad:
        text = replace_n_pattern(text, "<|endoftext|>", "{N} * <|endoftext|>")
    else:
        text = replace_n_pattern(text, "<|endoftext|>", "")
    text = replace_n_pattern(text, "<|video_pad|>", "{N} * <|video_pad|>")
    text = replace_n_pattern(text, "<|image_pad|>", "{N} * <|image_pad|>")
    return text.strip("\n")

def replace_n_pattern(text, pattern, replace_pattern):
    """
    替换字符串中连续出现的指定模式。

    参数:
        text (str): 原始字符串。
        pattern (str): 需要匹配的模式。
        replace_pattern (str): 替换模式，其中 `{N}` 会被替换为连续匹配的数量。

    返回:
        str: 替换后的字符串。
    """
    regex = re.compile(f'({re.escape(pattern)})+')
    
    def replace_match(match):
        matched_text = match.group()
        count = matched_text.count(pattern)
        return replace_pattern.replace('{N}', str(count))
    
    result = regex.sub(replace_match, text)
    return result


def collate_fn(examples, processor=None):
    # 1. 常规处理：生成 input_ids 和 基础 mask
    # 安全检查
    if processor is None:
        raise ValueError("严重错误: collate_fn 里的 processor 是 None！")   
    prompts_text = [
        processor.apply_chat_template(example["messages"], tokenize=False, add_generation_prompt=False)
        for example in examples
    ]
    messages = [deepcopy(example["messages"]) for example in examples]
    images, videos, video_kwargs = process_vision_info(messages, return_video_kwargs=True)
    
    batch = processor(
        text=deepcopy(prompts_text),
        images=images,
        videos=videos,
        fps=video_kwargs["fps"],
        padding=True, 
        return_tensors="pt", 
    )

    labels = batch["input_ids"].clone()
    
    # 2. 获取特殊 token ID
    tokenizer = processor.tokenizer
    im_start_id = tokenizer.convert_tokens_to_ids("<|im_start|>") # 151644
    assistant_id = tokenizer.convert_tokens_to_ids("assistant")   # 77091
    
    # 3. 基础 Mask：屏蔽所有 System 和 User 部分 (保留所有 Assistant)
    non_assistant_response_mask = compute_tool_response_mask(labels, im_start_id, assistant_id)
    labels[non_assistant_response_mask == 1] = -100 
    labels[labels == tokenizer.pad_token_id] = -100
    
    # 屏蔽视觉占位符
    video_token_id = tokenizer.convert_tokens_to_ids(processor.video_token)
    image_token_id = tokenizer.convert_tokens_to_ids(processor.image_token)
    labels[labels == video_token_id] = -100
    labels[labels == image_token_id] = -100

    
    # 使用你指定的长句子作为锚点。
    # 为了防止 Tokenizer 对长句编码产生细微差异（如空格），我们取前这句独特的话即可。
    anchor_text = "The interaction history above presents two distinct analysis phases"
    marker_ids = tokenizer.encode(anchor_text, add_special_tokens=False)
    marker_tensor = torch.tensor(marker_ids, dtype=torch.long)
    pat_len = len(marker_ids)
    for i, example in enumerate(examples):
        # 仅对 S3 样本执行此逻辑
        if example.get("split_source") == "s3":
            seq = batch["input_ids"][i]
            seq_len = seq.shape[0]
            
            # 1. 在 input_ids 中搜索 S3 User Prompt 的起始位置
            found_idx = -1
            
            # CPU 滑动窗口搜索
            # 优化：因为这句话出现在 prompt 的后半部分，其实可以从后往前搜，或者直接从头搜
            # 为了稳妥，我们还是从头搜，找到第一个匹配项（因为这句话是 S3 的开头，理论上只会出现一次）
            seq_cpu = seq.cpu() # 确保在 CPU 上操作
            for pos in range(seq_len - pat_len):
                if (seq_cpu[pos:pos+pat_len] == marker_tensor).all():
                    found_idx = pos
                    print(f"[S3] Found anchor at index {found_idx}")
                    break
            
            # 2. 如果找到了锚点，继续向后找 S3 Assistant 的开始
            if found_idx != -1:
                s3_assist_start = -1
                # 从锚点之后开始搜索 "<|im_start|>assistant"
                # 这必定是 S3 的 Assistant，因为 S1/S2 的 assistant 都在这个锚点之前
                for pos in range(found_idx, seq_len - 1):
                    if seq[pos] == im_start_id and seq[pos+1] == assistant_id:
                        s3_assist_start = pos
                        break
                
                # 3. 执行屏蔽：将 S3 Assistant 之前的所有 label 设为 -100
                if s3_assist_start != -1:
                    # 关键：这行代码会把 S1, S2 的 loss 全部抹去，只保留 S3
                    labels[i, :s3_assist_start] = -100
            else:
                # 如果没找到锚点，说明 Prompt 模板可能不对，或者被截断了
                # 这种情况下最好打印个 warning 或者跳过 masking (保留默认行为)
                # logger.warning(f"S3 Anchor not found in sample {i}")
                pass

    batch["labels"] = labels
    return batch









def main(script_args, training_args, model_args):
    # Set seed for reproducibility
    set_seed(training_args.seed)
    print("保存路径：", training_args.output_dir)
    ###############
    # Setup logging
    ###############
    logging.basicConfig(
        format="%(asctime)s - %(levelname)s - %(name)s - %(message)s",
        datefmt="%Y-%m-%d %H:%M:%S",
        handlers=[logging.StreamHandler(sys.stdout)],
    )
    log_level = training_args.get_process_log_level()
    logger.setLevel(log_level)
    datasets.utils.logging.set_verbosity(log_level)
    transformers.utils.logging.set_verbosity(log_level)
    transformers.utils.logging.enable_default_handler()
    transformers.utils.logging.enable_explicit_format()

    # Log on each process a small summary
    logger.warning(
        f"Process rank: {training_args.local_rank}, device: {training_args.device}, n_gpu: {training_args.n_gpu}"
        + f" distributed training: {bool(training_args.local_rank != -1)}, 16-bits training: {training_args.fp16}"
    )
    logger.info(f"Model parameters {model_args}")
    logger.info(f"Script parameters {script_args}")
    logger.info(f"Data parameters {training_args}")

    # Check for last checkpoint
    last_checkpoint = None
    if os.path.isdir(training_args.output_dir):
        last_checkpoint = get_last_checkpoint(training_args.output_dir)
    if last_checkpoint is not None and training_args.resume_from_checkpoint is None:
        logger.info(f"Checkpoint detected, resuming training at {last_checkpoint=}.")

    ################
    # Load datasets
    ################
    video_kwargs = {
        "total_pixels": script_args.total_video_tokens * 28 * 28,
        "min_pixels": script_args.min_per_frame_tokens * 28 * 28,
        "max_pixels": script_args.max_per_frame_tokens * 28 * 28,
        "max_frames": script_args.max_frames,
    }


    train_dataset = LazyVLDatasetSFT(script_args.train_data_path,
                                    script_args.video_folder,
                                    prompt_template=script_args.prompt_template,
                                    tool_name_list=script_args.tool_name_list,
                                    video_kwargs=video_kwargs)

    ################
    # Load tokenizer
    ################
    global processor
    if "vl" in model_args.model_name_or_path.lower():
        processor = AutoProcessor.from_pretrained(
            model_args.model_name_or_path, trust_remote_code=model_args.trust_remote_code
        )
        logger.info("Using AutoProcessor for vision-language model.")
    else:
        processor = AutoTokenizer.from_pretrained(
            model_args.model_name_or_path, trust_remote_code=model_args.trust_remote_code, use_fast=True
        )
        logger.info("Using AutoTokenizer for text-only model.")
    if hasattr(processor, "pad_token") and processor.pad_token is None:
        processor.pad_token = processor.eos_token
    elif hasattr(processor.tokenizer, "pad_token") and processor.tokenizer.pad_token is None:
        processor.tokenizer.pad_token = processor.tokenizer.eos_token

    ###################
    # Model init kwargs
    ###################
    logger.info("*** Initializing model kwargs ***")
    model_init_kwargs = training_args.model_init_kwargs or {}
    model_init_kwargs["use_cache"] = (
        False if training_args.gradient_checkpointing else model_init_kwargs.get("use_cache")
    )
    model_init_kwargs["attn_implementation"] = "flash_attention_2"
    model_init_kwargs['torch_dtype'] = torch.bfloat16
    from transformers import Qwen2_5_VLForConditionalGeneration
    model = Qwen2_5_VLForConditionalGeneration.from_pretrained(
        model_args.model_name_or_path, 
        **model_init_kwargs
    )
    ############################
    # Initialize the SFT Trainer
    ############################
    training_args.gradient_checkpointing_kwargs = {"use_reentrant": False}
    training_args.dataset_kwargs = {
        "skip_prepare_dataset": True,
    }
    from functools import partial
    my_collator = partial(collate_fn, processor=processor)
    training_args.remove_unused_columns = False
    trainer = SFTTrainer(
        model=model,
        args=training_args,
        train_dataset=train_dataset,
        processing_class=processor.tokenizer,
        data_collator=my_collator,
        peft_config=None
    )

    ###############
    # Training loop
    ###############
    logger.info("*** Train ***")
    checkpoint = None
    if training_args.resume_from_checkpoint is not None:
        checkpoint = training_args.resume_from_checkpoint
    elif last_checkpoint is not None:
        checkpoint = last_checkpoint
    train_result = trainer.train(resume_from_checkpoint=checkpoint)
    metrics = train_result.metrics
    metrics["train_samples"] = len(train_dataset)
    trainer.log_metrics("train", metrics)
    trainer.save_metrics("train", metrics)
    trainer.save_state()

    ##################################
    # Save model and create model card
    ##################################
    logger.info("*** Save model ***")
    trainer.save_model(training_args.output_dir)
    processor.save_pretrained(training_args.output_dir)
    logger.info(f"Model saved to {training_args.output_dir}")

    # Save everything else on main process
    kwargs = {
        "dataset_name": script_args.dataset_name,
        "tags": ["R1-V"],
    }
    if trainer.accelerator.is_main_process:
        trainer.create_model_card(**kwargs)
        # Restore k,v cache for fast inference
        trainer.model.config.use_cache = True
        trainer.model.config.save_pretrained(training_args.output_dir)
    #############
    # push to hub
    #############

    if training_args.push_to_hub:
        logger.info("Pushing to hub...")
        trainer.push_to_hub(**kwargs)
        processor.push_to_hub(training_args.hub_model_id)


if __name__ == "__main__":
    import multiprocessing
    try:
        multiprocessing.set_start_method('spawn', force=True)
    except RuntimeError:
        pass

    # 原有逻辑保持不变
    parser = TrlParser((SFTScriptArguments, SFTConfig, ModelConfig))
    script_args, training_args, model_args = parser.parse_args_and_config()
    main(script_args, training_args, model_args)

```

This is the sft.py.

**Comment by bb12346** [comment]
same problem!

**Comment by qgallouedec** [comment]
Hey @ShuimuChen-hyq, sorry for the slow reply, we're catching up on the backlog.

This is very hard for us to action as filed: the reproduction depends on a private dataset (700-frame videos) and custom modules (`LazyVLDatasetSFT`, your `collate_fn`, `time_r1.*`) that we can't run, and there's no `trl env` output.

A few pointers regardless. The `uncorrectable ECC error` in your title is a GPU hardware fault, not a TRL issue. When one rank hits it and drops off the bus, the others block on the next NCCL collective until the watchdog fires, which fits your "always the same fixed step" symptom (the same sample lands on the same GPU each run).
On the GRPO-vs-SFT difference: you pass `skip_prepare_dataset=True` and your own `collate_fn` that calls the processor with `padding=True` and no `max_length`, so TRL's truncation is fully bypassed and a 700-frame sample builds an essentially unbounded sequence. That's what OOMs under ZeRO-2 and stalls the parameter all-gather under ZeRO-3. I'd cap frames/tokens, truncate to a `max_length`, group samples by length to avoid rank imbalance, and run `nvidia-smi -q | grep -i ecc` to rule out the faulty GPU. 

Closing as not enough info to investigate
